File: programming-exercises/PHP/third.php
<?php

abstract class Component{
    abstract public function __toString();
}

class TextComponent extends Component{
    private $text;

    public function __toString(){
        return $this->text;
    }
    
    function __construct($text){
        $this->text = $text;
    }

    
}
class NumberComponent extends Component{
    private $number;
    function __construct($number){
        $this->number = $number;
    }

    function __toString(){
        return $this->number;
    }
}

class DateComponent extends Component{
    private $date;
    function __construct($date){
        $this->date = $date;
    }

    function __toString(){
        return $this->date;
    }
}

class CompositeComponent extends Component{
    private $list;

    function __toString(){
        $string = '';
        foreach ($this->list as $element){
            $string .= $element->__toString();
        }
        return $string;
    }

    function add(Component $element ){
        $this->list[] = clone $element;
    }

    function size(){
        return $this->list.count();
    }

    function modify($index, Component $element){
        $this->list[$index] = clone $element;
    }
}

$p = new CompositeComponent();

$p->add( new TextComponent("String. Number:"));
$p->add( new NumberComponent(1));
$p->add( new TextComponent("\n"));
$p->add( new TextComponent("Date:"));
$p->add( new DateComponent("7/7/12"));
$p->add( new TextComponent("\n"));
$d = new CompositeComponent();
$d->add( new TextComponent("Testing composite"));
$d->add( new TextComponent("\n"));
$d->add($p);
unset($p);
$d->add( new TextComponent("End testing composite"));

echo $d;

echo " new test\n";

$d->modify(0, new TextComponent("Beginning modified"));
echo $d;


?>

File: programming-exercises/Python/third.py
import copy

class Component:
  pass


class TextComponent(Component):
  
  def __init__(self, text):
    self.text = text

  def toString(self):
    return self.text

class NumberComponent(Component):
  
  def __init__(self, number):
    self.number = number

  def toString(self):
    return str(self.number)

class DateComponent(Component):
  
  def __init__(self, date):
    self.date = date

  def toString(self):
    return self.date

class CompositeComponent(Component):
  
  def __init__(self):
    self.listComponents = list()
  
  def add(self, component):
    if isinstance(component,Component):
      self.listComponents.append(copy.deepcopy(component))
    else:
      raise NameError("No component type")

  def size(self):
    return len(self.listComponents)

  def modify(self, index, component):
    if isinstance(component,Component):
      self.listComponents[index]=component
    else:
      raise NameError("No component type")

  def toString(self):
    string = ''
    for element in self.listComponents:
      string += element.toString()
    return string


p = CompositeComponent()
p.add(TextComponent("String. Number:"))
p.add(NumberComponent(1))
p.add(TextComponent("\n"))
p.add(TextComponent("Date:"))
p.add(DateComponent("02/02/02"))
p.add(TextComponent("\n"))
d = CompositeComponent()
d.add(TextComponent("Testing composite"))
d.add(TextComponent("\n"))
d.add(p)
del p
d.add(TextComponent("End testing composite"))

print d.toString()

print '----------------------'
d.modify(0,TextComponent("beginning modified"))

print d.toString()
